README
NAME
    Plack::App::GitHub::WebHook - GitHub WebHook receiver as Plack
    application

SYNOPSIS
  Basic usage
        use Plack::App::GitHub::WebHook;

        Plack::App::GitHub::WebHook->new(
            hook => sub {
                my $payload = shift;
                ...
            }
        )->to_app;

  Multiple task hooks
    A hook can consist of multiple tasks, given by an array reference. The
    tasks are called one by one until a task returns a false value.

        use Plack::App::GitHub::WebHook;
        use IPC::Run3;

        Plack::App::GitHub::WebHook->new(
            hook => [
                sub { $_[0]->{repository}{name} eq 'foo' }, # filter
                { Filter => { repository_name => 'foo' } }, # equivalent filter
                sub { my ($payload) = @_; ...  }, # some action
                sub { run3 \@cmd ... }, # some more action
            ]
        )->to_app;

  Access restriction
    By default access is restricted to known GitHub WebHook IPs.

        Plack::App::GitHub::WebHook->new(
            hook => sub { ... },
            access => [
                allow => "204.232.175.64/27",
                allow => "192.30.252.0/22",
                deny  => 'all'
            ]
        )->to_app;

        # this is equivalent to
        use Plack::Builder;
        builder {
            mount 'notify' => builder {
                enable 'Access', rules => [
                    allow => "204.232.175.64/27",
                    allow => "192.30.252.0/22",
                    deny  => 'all'
                ]
                Plack::App::GitHub::WebHook->new(
                    hook => sub { ... }
                );
            }
        };

  Synchronize with a GitHub repository
    The following application automatically pulls the master branch of a
    GitHub repository into a local working directory $work_tree.

        use Git::Repository
        use Plack::App::GitHub::WebHook;

        Plack::App::GitHub::WebHook->new(
            events => ['pull'],
            safe => 1,
            hook => [
                sub { $_[0]->{ref} eq 'refs/heads/master' },
                sub {
                    if ( -d "$work_tree/.git") {
                        Git::Repository->new( work_tree => $work_tree )
                                       ->run(qw(pull origin master));
                    } else {
                        my $origin = $_[0]->{repository}->{clone_url};
                        Git::Repository->run( 'clone', $origin, $work_tree );
                    }
                    1;
                },
                # sub { ...optional action after each pull... } 
            ],
        )->to_app;

DESCRIPTION
    This PSGI application receives HTTP POST requests with body parameter
    "payload" set to a JSON object. The default use case is to receive
    GitHub WebHooks <http://developer.github.com/webhooks/>, for instance
    PushEvents
    <http://developer.github.com/v3/activity/events/types/#pushevent>.

    The response of a HTTP request to this application is one of:

    HTTP 403 Forbidden
        If access was not granted (for instance because it did not origin
        from GitHub).

    HTTP 405 Method Not Allowed
        If the request was no HTTP POST.

    HTTP 400 Bad Request
        If the payload was no well-formed JSON or the "X-GitHub-Event"
        header did not match configured events.

    HTTP 200 OK
        Otherwise, if the hook was called and returned a true value.

    HTTP 202 Accepted
        Otherwise, if the hook was called and returned a false value.

    This module requires at least Perl 5.10.

CONFIGURATION
    hook
        A code reference or an array of code references with tasks that are
        executed on an incoming webhook. Each task gets passed the encoded
        payload. If the task returns a true value, next the task is called
        or HTTP status code 200 is returned. Information can be passed from
        one task to the next by modifying the payload.

        If a task returns a false value or if no task was given, HTTP status
        code 202 is returned immediately. This mechanism can be used for
        conditional hooks or to detect hooks that were called successfully
        but failed to execute for some reason.

    safe
        Wrap all hook tasks in "eval { ... }" blocks to catch exceptions. A
        dying task in safe mode is equivalent to a task that returns a false
        value.

    access
        Access restrictions, as passed to Plack::Middleware::Access. See
        SYNOPSIS for the default value. A recent list of official GitHub
        WebHook IPs is vailable at <https://api.github.com/meta>. One should
        only set the access value on instantiation, or manually call
        "prepare_app" after modification.

    events
        A list of event types
        <http://developer.github.com/v3/activity/events/types/> expected to
        be send with the "X-GitHub-Event" header (e.g. "['pull']").

SEE ALSO
    *   GitHub WebHooks are documented at
        <http://developer.github.com/webhooks/>.

    *   WWW::GitHub::PostReceiveHook uses Web::Simple to receive GitHub web
        hooks. A listener as exemplified by the module can also be created
        like this:

            use Plack::App::GitHub::WebHook;
            use Plack::Builder;
            build {
                mount '/myProject' => 
                    Plack::App::GitHub::WebHook->new(
                        hook => sub { my $payload = shift; }
                    );
                mount '/myOtherProject' => 
                    Plack::App::GitHub::WebHook->new(
                        hook => sub { run3 \@cmd ... }
                    );
            };

    *   Net::GitHub and Pithub provide access to GitHub APIs.

    *   App::GitHubWebhooks2Ikachan is an application that also receives
        GitHub WebHooks.

AUTHOR
    Jakob Voss "<jakob.voss@gbv.de>"

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
# Change file for Plack::App::GitHub::WebHook

0.5  2014-09-30 14:52:44 CEST
    - New options: events and safe
    - Migrated to Dist::Milla

0.4 2014-01-09
    - Support multiple hooks, chained

0.3 2014-01-02
    - Additional return code 202
    - Minor documentation fixes (thanks to Tomáš Znamenáček)

0.2 2013-07-30
    - First public version at CPAN


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.008.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "GitHub WebHook receiver as Plack application",
  "AUTHOR" => "Jakob Vo\x{df}",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Plack-App-GitHub-WebHook",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Plack::App::GitHub::WebHook",
  "PREREQ_PM" => {
    "JSON" => "2.0",
    "Plack" => "1.0",
    "Plack::Middleware::Access" => "0.3"
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.4",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "JSON" => "2.0",
  "Plack" => "1.0",
  "Plack::Middleware::Access" => "0.3"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Plack/App/GitHub/WebHook.pm
use strict;
package Plack::App::GitHub::WebHook;
#ABSTRACT: GitHub WebHook receiver as Plack application
our $VERSION = '0.4'; #VERSION
use v5.10;
use JSON qw(decode_json);

use parent 'Plack::Component';
use Plack::Util::Accessor qw(hook access app);
use Plack::Request;
use Plack::Middleware::Access;
use Carp qw(croak);

sub prepare_app {
    my $self = shift;

    if ($self->hook) {
        if ( (ref $self->hook // '') eq 'CODE' ) {
            $self->hook( [ $self->hook ] );
        } elsif ( (ref $self->hook // '') ne 'ARRAY' ) {
            croak "hook must be a CODEREF or ARRAYREF";
        }
    } else {
        $self->hook([]);
    }

    $self->access([
        allow => "204.232.175.64/27",
        allow => "192.30.252.0/22",
        deny  => "all"
    ]) unless $self->access;

    $self->app(
        Plack::Middleware::Access->wrap(
            sub { $self->call_granted(shift) },
            rules => $self->access
        )
    );

    $self->init;
}

sub init { }

sub call {
    my ($self, $env) = @_;
    $self->app->($env);
}

sub call_granted {
    my ($self, $env) = @_;

    if ( $env->{REQUEST_METHOD} ne 'POST' ) {
        return [405,['Content-Type'=>'text/plain','Content-Length'=>18],['Method Not Allowed']];
    }

    my $req = Plack::Request->new($env);

    my $json = eval { decode_json $req->body_parameters->{payload} };

    if (!$json) {
        return [400,['Content-Type'=>'text/plain','Content-Length'=>11],['Bad Request']];
    }

    if ( $self->receive($json) ) {
        return [200,['Content-Type'=>'text/plain','Content-Length'=>2],['OK']];
    } else {
        return [202,['Content-Type'=>'text/plain','Content-Length'=>8],['Accepted']];
    }
}

sub receive {
    my ($self, $payload) = @_;

    my $ok;
    foreach my $hook (@{$self->{hook}}) {
        return unless $hook->($payload);
        $ok++;
    } 
    return $ok;
}


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Plack::App::GitHub::WebHook - GitHub WebHook receiver as Plack application

=head1 VERSION

version 0.4

=head1 SYNOPSIS

    use Plack::App::GitHub::WebHook;

    Plack::App::GitHub::WebHook->new(
        hook => sub {
            my $payload = shift;
            ...
        }
    )->to_app;

=head2 Multiple hooks

If multiple hooks are provided, they get called one by one until
a hook returns a false value.

    use Plack::App::GitHub::WebHook;
    use IPC::Run3;

    Plack::App::GitHub::WebHook->new(
        hook => [
            sub { $_[0]->{repository}{name} eq 'foo' }, # filter
            sub { my ($payload) = @_; ...  }, # some action
            sub { run3 \@cmd ... }, # some more action
        ]
    )->to_app;

=head2 Access restriction    

By default access is restricted to known GitHub WebHook IPs.

    Plack::App::GitHub::WebHook->new(
        hook => sub { ... },
        access => [
            allow => "204.232.175.64/27",
            allow => "192.30.252.0/22",
            deny  => 'all'
        ]
    )->to_app;

    # this is equivalent to
    use Plack::Builder;
    builder {
        mount 'notify' => builder {
            enable 'Access', rules => [
                allow => "204.232.175.64/27",
                allow => "192.30.252.0/22",
                deny  => 'all'
            ]
            Plack::App::GitHub::WebHook->new(
                hook => sub { ... }
            );
        }
    };

=head1 DESCRIPTION

This L<PSGI> application receives HTTP POST requests with body parameter
C<payload> set to a JSON object. The default use case is to receive
L<GitHub WebHooks|https://help.github.com/articles/post-receive-hooks>.

The response of a HTTP request to this application is one of:

=over 4

=item HTTP 403 Forbidden

If access was not granted (for instance because it did not origin from GitHub).

=item HTTP 405 Method Not Allowed

If the request was no HTTP POST.

=item HTTP 400 Bad Request

If the payload was no well-formed JSON. A later version of this module may add
further validation.

=item HTTP 200 OK

Otherwise, if the hook was called and returned a true value.

=item HTTP 202 Accepted

Otherwise, if the hook was called and returned a false value.

=back

This module requires at least Perl 5.10.

=head1 CONFIGURATION

=over 4

=item hook

A code reference or an array reference of code references with multiple tasks.
Each task gets passed the encoded payload. If the task returns a true value,
next the task is called or HTTP status code 200 is returned. Information can be
passed from one task to the next by modifying the payload. 

If a task fails or no task was given, HTTP status code 202 is returned
immediately. This mechanism can be used for conditional hooks or to detect
hooks that were called successfully but failed to execute for some reason.

=item access

Access restrictions, as passed to L<Plack::Middleware::Access>. See SYNOPSIS
for the default value. A recent list of official GitHub WebHook IPs is vailable
at L<https://api.github.com/meta>. One should only set the access value on
instantiation, or manually call C<prepare_app> after modification.

=back

=head1 SEE ALSO

L<WWW::GitHub::PostReceiveHook> uses L<Web::Simple> to receive GitHub web
hooks. L<Net::GitHub> and L<Pithub> provide access to GitHub APIs.

=head1 AUTHOR

Jakob Voß

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
